File: deepeval/metrics/tool_correctness/tool_correctness.py
import asyncio
from typing import List, Union, Dict, Optional

from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.metrics.utils import (
    construct_verbose_logs,
    check_llm_test_case_params,
    initialize_model,
    trimAndLoadJson,
)
from deepeval.test_case import (
    LLMTestCase,
    LLMTestCaseParams,
    ToolCallParams,
    ToolCall,
    MCPMetaData,
    MCPToolCall,
    MCPResourceCall,
    MCPPromptCall,
)
from deepeval.utils import get_or_create_event_loop
from deepeval.metrics import BaseMetric
from deepeval.models import DeepEvalBaseLLM
from .template import ToolCorrectnessTemplate
from .schema import ToolScore


class ToolCorrectnessMetric(BaseMetric):

    _required_params: List[LLMTestCaseParams] = [
        LLMTestCaseParams.INPUT,
        LLMTestCaseParams.ACTUAL_OUTPUT,
    ]

    def __init__(
        self,
        threshold: float = 0.5,
        evaluation_params: List[ToolCallParams] = [],
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        include_reason: bool = True,
        strict_mode: bool = False,
        async_mode: bool = True,
        verbose_mode: bool = False,
        should_exact_match: bool = False,
        should_consider_ordering: bool = False,
    ):
        self.threshold = 1 if strict_mode else threshold
        self.include_reason = include_reason
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()
        self.strict_mode = strict_mode
        self.async_mode = async_mode
        self.verbose_mode = verbose_mode
        self.evaluation_params: List[ToolCallParams] = evaluation_params
        self.should_exact_match = should_exact_match
        self.should_consider_ordering = should_consider_ordering

    def measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:
        if test_case.tools_called is not None:
            self._required_params.append(LLMTestCaseParams.TOOLS_CALLED)
            self._required_params.append(LLMTestCaseParams.EXPECTED_TOOLS)
        if (
            test_case.mcp_tools_called is not None
            or test_case.mcp_resources_called is not None
            or test_case.mcp_prompts_called is not None
        ):
            self._required_params.append(LLMTestCaseParams.MCP_DATA)
        check_llm_test_case_params(test_case, self._required_params, self)

        self.test_case = test_case
        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                    )
                )
            if test_case.tools_called:
                self.tools_called: List[ToolCall] = test_case.tools_called
                self.expected_tools: List[ToolCall] = test_case.expected_tools
                self.tools_score = self._calculate_score()
                self.tools_reason = self._generate_reason()
            if test_case.mcp_data:
                self.mcp_data: List[MCPMetaData] = test_case.mcp_data
                self.mcp_tools_called: Optional[List[MCPToolCall]] = (
                    test_case.mcp_tools_called
                )
                self.mcp_resources_called: Optional[List[MCPResourceCall]] = (
                    test_case.mcp_resources_called
                )
                self.mcp_prompts_called: Optional[List[MCPPromptCall]] = (
                    test_case.mcp_prompts_called
                )
                available_primitives, primitives_used = (
                    self._get_mcp_interaction_text()
                )
                self.mcp_interaction = (
                    available_primitives + "\n" + primitives_used
                )
                self.mcp_score: ToolScore = self._get_mcp_score(
                    test_case, available_primitives, primitives_used
                )

            self.score = self._calculate_final_score(test_case)
            self.reason = self._get_final_reason(test_case)
            self.success = self.score >= self.threshold
            expected_tools_formatted = (
                "Expected Tools:\n[\n"
                + ",\n".join(
                    self.indent_multiline_string(
                        repr(tool_call), indent_level=4
                    )
                    for tool_call in self.expected_tools
                )
                + "\n]"
            )
            tools_called_formatted = (
                "Tools Called:\n[\n"
                + ",\n".join(
                    self.indent_multiline_string(
                        repr(tool_call), indent_level=4
                    )
                    for tool_call in self.tools_called
                )
                + "\n]"
            )
            steps = [
                f"{expected_tools_formatted}",
                f"{tools_called_formatted}",
            ]
            if self.mcp_interaction:
                steps.append(self.mcp_interaction)
            steps.append(f"Score: {self.score}\nReason: {self.reason}")
            self.verbose_logs = construct_verbose_logs(self, steps=steps)
            return self.score

    async def a_measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:
        check_llm_test_case_params(test_case, self._required_params, self)

        self.test_case = test_case
        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if test_case.tools_called:
                self.tools_called: List[ToolCall] = test_case.tools_called
                self.expected_tools: List[ToolCall] = test_case.expected_tools
                self.tools_score = self._calculate_score()
                self.tools_reason = self._generate_reason()
            if test_case.mcp_data:
                self.mcp_data: List[MCPMetaData] = test_case.mcp_data
                self.mcp_tools_called: Optional[List[MCPToolCall]] = (
                    test_case.mcp_tools_called
                )
                self.mcp_resources_called: Optional[List[MCPResourceCall]] = (
                    test_case.mcp_resources_called
                )
                self.mcp_prompts_called: Optional[List[MCPPromptCall]] = (
                    test_case.mcp_prompts_called
                )
                available_primitives, primitives_used = (
                    self._get_mcp_interaction_text()
                )
                self.mcp_interaction = (
                    available_primitives + "\n" + primitives_used
                )
                self.mcp_score: ToolScore = await self._a_get_mcp_score(
                    test_case, available_primitives, primitives_used
                )

            self.score = self._calculate_final_score(test_case)
            self.reason = self._get_final_reason(test_case)
            self.success = self.score >= self.threshold
            steps = []
            if test_case.tools_called:
                expected_tools_formatted = (
                    "Expected Tools:\n[\n"
                    + ",\n".join(
                        self.indent_multiline_string(
                            repr(tool_call), indent_level=4
                        )
                        for tool_call in self.expected_tools
                    )
                    + "\n]"
                )
                tools_called_formatted = (
                    "Tools Called:\n[\n"
                    + ",\n".join(
                        self.indent_multiline_string(
                            repr(tool_call), indent_level=4
                        )
                        for tool_call in self.tools_called
                    )
                    + "\n]"
                )
                steps.append(f"{expected_tools_formatted}")
                steps.append(f"{tools_called_formatted}")
            if test_case.mcp_data:
                steps.append(self.mcp_interaction)
            steps.append(f"Score: {self.score}\nReason: {self.reason}")
            self.verbose_logs = construct_verbose_logs(self, steps=steps)
            return self.score

    ##################################################
    ### Tool Correctness (Tool) ######################
    ##################################################

    def _generate_reason(self):
        tools_called_names = [
            tool_called.name for tool_called in self.tools_called
        ]
        expected_tools_names = [
            expected_tool.name for expected_tool in self.expected_tools
        ]

        if self.should_exact_match:
            return f"{'Exact match' if self._calculate_exact_match_score() else 'Not an exact match'}: expected {expected_tools_names}, called {tools_called_names}. See details above."

        elif self.should_consider_ordering:
            lcs, weighted_length = self._compute_weighted_lcs()
            score = weighted_length / len(expected_tools_names)
            missing = set(expected_tools_names) - set(tools_called_names)
            out_of_order = set(expected_tools_names) - set(
                [tool.name for tool in lcs]
            )
            if score == 1:
                return f"Correct ordering: all expected tools {expected_tools_names} were called in the correct order."
            else:
                issues = []
                if missing:
                    issues.append(f"missing tools {list(missing)}")
                if out_of_order:
                    issues.append(f"out-of-order tools {list(out_of_order)}")
                return f"Incorrect tool usage: {' and '.join(issues)}; expected {expected_tools_names}, called {tools_called_names}. See more details above."
        else:
            used_expected = set(self.tools_called).intersection(
                set(self.expected_tools)
            )
            missing = set(self.expected_tools) - used_expected
            if self._calculate_non_exact_match_score() == 1:
                return f"All expected tools {expected_tools_names} were called (order not considered)."
            else:
                return f"Incomplete tool usage: missing tools {list(missing)}; expected {expected_tools_names}, called {tools_called_names}. See more details above."

    ##################################################
    ### Score Helper Functions #######################
    ##################################################

    # Get mcp interaction text
    def _get_mcp_interaction_text(self) -> tuple[str, str]:
        for mcp_data in self.mcp_data:
            available_primitives = f"MCP Server {mcp_data.server_name}\n"
            available_primitives += (
                (
                    "\nAvailable Tools:\n[\n"
                    + ",\n".join(
                        self.indent_multiline_string(repr(tool), indent_level=4)
                        for tool in mcp_data.available_tools
                    )
                    + "\n]"
                )
                if mcp_data.available_tools
                else ""
            )
            available_primitives += (
                (
                    "\nAvailable Resources:\n[\n"
                    + ",\n".join(
                        self.indent_multiline_string(
                            repr(resource), indent_level=4
                        )
                        for resource in mcp_data.available_resources
                    )
                    + "\n]"
                )
                if mcp_data.available_resources
                else ""
            )
            available_primitives += (
                (
                    "\nAvailable Prompts:\n[\n"
                    + ",\n".join(
                        self.indent_multiline_string(
                            repr(prompt), indent_level=4
                        )
                        for prompt in mcp_data.available_prompts
                    )
                    + "\n]"
                )
                if mcp_data.available_prompts
                else ""
            )
        primitives_used = "MCP Primitives Used: \n"
        primitives_used += (
            (
                "\nMCP Tools Called:\n[\n"
                + ",\n".join(
                    self.indent_multiline_string(
                        repr(mcp_tool_call), indent_level=4
                    )
                    for mcp_tool_call in self.mcp_tools_called
                )
                + "\n]"
            )
            if self.mcp_tools_called
            else ""
        )
        primitives_used += (
            (
                "\nMCP Resources Called:\n[\n"
                + ",\n".join(
                    self.indent_multiline_string(
                        repr(mcp_resource_call), indent_level=4
                    )
                    for mcp_resource_call in self.mcp_resources_called
                )
                + "\n]"
            )
            if self.mcp_resources_called
            else ""
        )
        primitives_used += (
            (
                "\nMCP Prompts Called:\n[\n"
                + ",\n".join(
                    self.indent_multiline_string(
                        repr(mcp_prompt_call), indent_level=4
                    )
                    for mcp_prompt_call in self.mcp_prompts_called
                )
                + "\n]"
            )
            if self.mcp_prompts_called
            else ""
        )

        return available_primitives, primitives_used

    # Get MCP Score
    def _get_mcp_score(
        self,
        test_case: LLMTestCase,
        available_primitives: str,
        primitives_used: str,
    ):
        prompt = ToolCorrectnessTemplate.get_tool_correctness_prompt(
            test_case, available_primitives, primitives_used
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=ToolScore)
            self.evaluation_cost += cost
            return res
        else:
            try:
                res: ToolScore = self.model.generate(prompt, schema=ToolScore)
                return res
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return ToolScore(**data)

    async def _a_get_mcp_score(
        self,
        test_case: LLMTestCase,
        available_primitives: str,
        primitives_used: str,
    ):
        prompt = ToolCorrectnessTemplate.get_tool_correctness_prompt(
            test_case, available_primitives, primitives_used
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=ToolScore)
            self.evaluation_cost += cost
            return res
        else:
            try:
                res: ToolScore = await self.model.a_generate(
                    prompt, schema=ToolScore
                )
                return res
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return ToolScore(**data)

    # Calculate Final Score
    def _calculate_final_score(self, test_case: LLMTestCase):
        if test_case.tools_called and test_case.mcp_data:
            return (self.tools_score + self.mcp_score.score) / 2
        elif test_case.mcp_data:
            return self.mcp_score.score
        else:
            return self.tools_score

    def _get_final_reason(self, test_case: LLMTestCase):
        if test_case.tools_called and test_case.mcp_data:
            reasons = (
                f"[\n"
                f"\t{self.tools_reason}, \n"
                f"\t{self.mcp_score.reason} \n"
                f"]"
            )
            return reasons
        elif test_case.mcp_data:
            return self.mcp_score.reason
        else:
            return self.tools_reason

    # Calculate score
    def _calculate_score(self):
        if self.should_exact_match:
            score = self._calculate_exact_match_score()
        elif self.should_consider_ordering:
            _, weighted_length = self._compute_weighted_lcs()
            score = weighted_length / len(self.expected_tools)
        else:
            score = self._calculate_non_exact_match_score()
        return 0 if self.strict_mode and score < self.threshold else score

    # Exact matching score
    def _calculate_exact_match_score(self):
        if len(self.tools_called) != len(self.expected_tools):
            return 0.0
        for i in range(len(self.tools_called)):
            if self.tools_called[i].name != self.expected_tools[i].name:
                return 0.0
            if ToolCallParams.INPUT_PARAMETERS in self.evaluation_params:
                if (
                    self.tools_called[i].input_parameters
                    != self.expected_tools[i].input_parameters
                ):
                    return 0.0
            if ToolCallParams.OUTPUT in self.evaluation_params:
                if self.tools_called[i].output != self.expected_tools[i].output:
                    return 0.0
        return 1.0

    # Non exact matching score
    def _calculate_non_exact_match_score(self):
        total_score = 0.0
        matched_called_tools = set()
        for expected_tool in self.expected_tools:
            best_score = 0.0
            for called_tool in self.tools_called:
                if called_tool in matched_called_tools:
                    continue
                if expected_tool.name == called_tool.name:
                    match_score = 1.0
                    if (
                        ToolCallParams.INPUT_PARAMETERS
                        in self.evaluation_params
                    ):
                        match_score *= self._compare_dicts(
                            expected_tool.input_parameters,
                            called_tool.input_parameters,
                        )
                    if (
                        ToolCallParams.OUTPUT in self.evaluation_params
                        and expected_tool.output != called_tool.output
                    ):
                        match_score = 0.0
                    if match_score > best_score:
                        best_score = match_score
                        best_called_tool = called_tool
            if best_score > 0:
                total_score += best_score
                matched_called_tools.add(best_called_tool)
        return (
            total_score / len(self.expected_tools)
            if self.expected_tools
            else 0.0
        )

    # Consider ordering score
    def _compute_weighted_lcs(self):
        m, n = len(self.expected_tools), len(self.tools_called)
        dp = [[0.0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                expected_tool, called_tool = (
                    self.expected_tools[i - 1],
                    self.tools_called[j - 1],
                )
                if expected_tool.name != called_tool.name:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
                    continue
                score = 1.0
                if ToolCallParams.INPUT_PARAMETERS in self.evaluation_params:
                    score *= self._compare_dicts(
                        expected_tool.input_parameters,
                        called_tool.input_parameters,
                    )
                if (
                    ToolCallParams.OUTPUT in self.evaluation_params
                    and expected_tool.output != called_tool.output
                ):
                    score = 0.0
                dp[i][j] = max(
                    dp[i - 1][j],
                    dp[i][j - 1],
                    dp[i - 1][j - 1] + score if score > 0 else 0,
                )
        i, j, total_score = m, n, 0.0
        lcs = []
        while i > 0 and j > 0:
            if dp[i][j] == dp[i - 1][j]:
                i -= 1
            elif dp[i][j] == dp[i][j - 1]:
                j -= 1
            else:
                lcs.append(self.expected_tools[i - 1])
                total_score += dp[i][j] - dp[i - 1][j - 1]
                i, j = i - 1, j - 1
        return lcs[::-1], total_score

    # For matching input parameters
    def _compare_dicts(self, dict1: Dict, dict2: Dict):
        if dict1 == dict2:
            return 1.0
        if self.should_exact_match:
            return 1.0 if dict1 == dict2 else 0.0
        match_score = 0
        matched_keys = set(dict1.keys()).intersection(set(dict2.keys()))
        total_keys = set(dict1.keys()).union(set(dict2.keys()))
        for key in matched_keys:
            if dict1[key] == dict2[key]:
                match_score += 1 / len(total_keys)
            elif isinstance(dict1[key], dict) and isinstance(dict2[key], dict):
                match_score += self._compare_dicts(
                    dict1[key], dict2[key]
                ) / len(total_keys)
        return match_score

    ##################################################
    ### Others #######################################
    ##################################################

    def is_successful(self) -> bool:
        try:
            self.success = self.score >= self.threshold
        except:
            self.success = False
        return self.success

    @property
    def __name__(self):
        return "Tool Correctness"

    def indent_multiline_string(self, s, indent_level=4):
        indent = " " * indent_level
        return "\n".join(f"{indent}{line}" for line in s.splitlines())

File: typescript/src/metrics/tool-correctness/tool-correctness.ts
import { BaseMetric } from "../base-metrics";
import {
  LLMTestCase,
  SingleTurnParams,
  ToolCallParams,
  ToolCall,
} from "../../test-case";
import { DeepEvalBaseLLM } from "../../models";
import { resolveTemplate } from "../../templates";
import {
  initializeModel,
  generateWithSchema,
  checkSingleTurnParams,
  constructVerboseLogs,
  printToolsCalled,
} from "../utils";
import { ToolSelectionScoreSchema, type ToolSelectionScore } from "./schema";

const TEMPLATE_CLASS = "ToolCorrectnessMetric";

/** Order-insensitive deep equality (matches Python `==` on dicts/values). */
function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a == null || b == null) return a === b;
  if (typeof a !== typeof b) return false;
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((x, i) => deepEqual(x, b[i]));
  }
  if (typeof a === "object" && typeof b === "object") {
    const ka = Object.keys(a as object);
    const kb = Object.keys(b as object);
    if (ka.length !== kb.length) return false;
    return ka.every((k) =>
      deepEqual(
        (a as Record<string, unknown>)[k],
        (b as Record<string, unknown>)[k],
      ),
    );
  }
  return false;
}

/** ToolCall equality: same name, input parameters, and output (Python `__eq__`). */
function toolCallEquals(a: ToolCall, b: ToolCall): boolean {
  return (
    a.name === b.name &&
    deepEqual(a.inputParameters, b.inputParameters) &&
    deepEqual(a.output, b.output)
  );
}

/** Dedup a list of names, preserving Python `set()`-style membership. */
function uniqueMissing(expected: string[], called: string[]): string[] {
  const calledSet = new Set(called);
  return [...new Set(expected)].filter((n) => !calledSet.has(n));
}

export interface ToolCorrectnessMetricOptions {
  /** If provided (and non-empty), an LLM also judges tool *selection*. */
  availableTools?: ToolCall[];
  threshold?: number;
  /** Which `ToolCall` fields to compare (input parameters / output). */
  evaluationParams?: ToolCallParams[];
  model?: DeepEvalBaseLLM | string;
  includeReason?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  shouldExactMatch?: boolean;
  shouldConsiderOrdering?: boolean;
}

/**
 * Tool Correctness — were the right tools called (vs. `expectedTools`)?
 * Deterministic tool-calling score (exact / ordered / unordered), combined via
 * `min` with an optional LLM tool-*selection* score when `availableTools` is set.
 */
export class ToolCorrectnessMetric extends BaseMetric {
  toolsCalled: ToolCall[] = [];
  expectedTools: ToolCall[] = [];
  private readonly availableTools?: ToolCall[];
  private readonly evaluationParams: ToolCallParams[];
  private readonly shouldExactMatch: boolean;
  private readonly shouldConsiderOrdering: boolean;

  constructor(options: ToolCorrectnessMetricOptions = {}) {
    const strictMode = options.strictMode ?? false;
    super(strictMode ? 1 : (options.threshold ?? 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
    });
    this.requiredParams = [
      SingleTurnParams.INPUT,
      SingleTurnParams.TOOLS_CALLED,
      SingleTurnParams.EXPECTED_TOOLS,
    ];
    this.availableTools = options.availableTools;
    this.evaluationParams = options.evaluationParams ?? [];
    this.shouldExactMatch = options.shouldExactMatch ?? false;
    this.shouldConsiderOrdering = options.shouldConsiderOrdering ?? false;
    const { model, usingNativeModel } = initializeModel(options.model);
    this.model = model;
    this.usingNativeModel = usingNativeModel;
    this.evaluationModel = this.model.getModelName();
  }

  async measure(testCase: LLMTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      checkSingleTurnParams(testCase, this.requiredParams, this);
      this.evaluationCost = this.usingNativeModel ? 0 : undefined;

      this.toolsCalled = testCase.toolsCalled ?? [];
      this.expectedTools = testCase.expectedTools ?? [];

      const toolCallingScore = this.calculateScore();
      const toolSelectionScore: ToolSelectionScore =
        this.availableTools && this.availableTools.length > 0
          ? await this.getToolSelectionScore(testCase.input)
          : {
              score: 1,
              reason:
                "No available tools were provided to assess tool selection criteria",
            };

      const combined = Math.min(toolCallingScore, toolSelectionScore.score);
      this.score =
        this.strictMode && combined < this.threshold ? 0 : combined;
      this.reason = this.constructFinalReason(
        this.generateReason(),
        toolSelectionScore.reason,
      );
      this.success = this.score >= this.threshold;

      this.verboseLogs = constructVerboseLogs(this, [
        `Expected Tools:\n${printToolsCalled(this.expectedTools)}`,
        `Tools Called:\n${printToolsCalled(this.toolsCalled)}`,
        `Available Tools:\n${this.availableTools ? printToolsCalled(this.availableTools) : "[]"}`,
        `Tool Selection Score: ${toolSelectionScore.score}`,
        `Tool Selection Reason: ${toolSelectionScore.reason}`,
        `Final Score: ${this.score}\nFinal Reason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  private async getToolSelectionScore(
    userInput: string,
  ): Promise<ToolSelectionScore> {
    const prompt = resolveTemplate("metrics", TEMPLATE_CLASS, "get_tool_selection_score", {
      user_input: userInput,
      tools_called: printToolsCalled(this.toolsCalled),
      available_tools: printToolsCalled(this.availableTools ?? []),
    });
    return generateWithSchema(this, prompt, ToolSelectionScoreSchema);
  }

  // --- scoring ---

  private calculateScore(): number {
    let score: number;
    if (this.shouldExactMatch) {
      score = this.calculateExactMatchScore();
    } else if (this.shouldConsiderOrdering) {
      const { weightedLength } = this.computeWeightedLcs();
      if (this.toolsCalled.length === 0 && this.expectedTools.length === 0) {
        score = 1;
      } else if (this.expectedTools.length === 0) {
        score = 0;
      } else {
        score = weightedLength / this.expectedTools.length;
      }
    } else {
      score = this.calculateNonExactMatchScore();
    }
    return this.strictMode && score < this.threshold ? 0 : score;
  }

  private calculateExactMatchScore(): number {
    if (this.toolsCalled.length !== this.expectedTools.length) return 0;
    if (this.expectedTools.length === 0) return 1;
    for (let i = 0; i < this.toolsCalled.length; i++) {
      const called = this.toolsCalled[i];
      const expected = this.expectedTools[i];
      if (called.name !== expected.name) return 0;
      if (
        this.evaluationParams.includes(ToolCallParams.INPUT_PARAMETERS) &&
        !deepEqual(called.inputParameters, expected.inputParameters)
      ) {
        return 0;
      }
      if (
        this.evaluationParams.includes(ToolCallParams.OUTPUT) &&
        !deepEqual(called.output, expected.output)
      ) {
        return 0;
      }
    }
    return 1;
  }

  private calculateNonExactMatchScore(): number {
    let totalScore = 0;
    const matchedCalled = new Set<number>();
    for (const expected of this.expectedTools) {
      let bestScore = 0;
      let bestIdx = -1;
      for (let j = 0; j < this.toolsCalled.length; j++) {
        if (matchedCalled.has(j)) continue;
        const called = this.toolsCalled[j];
        if (expected.name !== called.name) continue;
        let matchScore = 1;
        if (this.evaluationParams.includes(ToolCallParams.INPUT_PARAMETERS)) {
          matchScore *= this.compareDicts(
            expected.inputParameters ?? {},
            called.inputParameters ?? {},
          );
        }
        if (
          this.evaluationParams.includes(ToolCallParams.OUTPUT) &&
          !deepEqual(expected.output, called.output)
        ) {
          matchScore = 0;
        }
        if (matchScore > bestScore) {
          bestScore = matchScore;
          bestIdx = j;
        }
      }
      if (bestScore > 0) {
        totalScore += bestScore;
        matchedCalled.add(bestIdx);
      }
    }
    if (this.expectedTools.length === 0 && this.toolsCalled.length === 0)
      return 1;
    if (this.expectedTools.length === 0) return 0;
    return totalScore / this.expectedTools.length;
  }

  private computeWeightedLcs(): { lcs: ToolCall[]; weightedLength: number } {
    const expected = this.expectedTools;
    const called = this.toolsCalled;
    const m = expected.length;
    const n = called.length;
    const dp: number[][] = Array.from({ length: m + 1 }, () =>
      new Array(n + 1).fill(0),
    );
    for (let i = 1; i <= m; i++) {
      for (let j = 1; j <= n; j++) {
        const e = expected[i - 1];
        const c = called[j - 1];
        if (e.name !== c.name) {
          dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
          continue;
        }
        let score = 1;
        if (this.evaluationParams.includes(ToolCallParams.INPUT_PARAMETERS)) {
          score *= this.compareDicts(
            e.inputParameters ?? {},
            c.inputParameters ?? {},
          );
        }
        if (
          this.evaluationParams.includes(ToolCallParams.OUTPUT) &&
          !deepEqual(e.output, c.output)
        ) {
          score = 0;
        }
        dp[i][j] = Math.max(
          dp[i - 1][j],
          dp[i][j - 1],
          score > 0 ? dp[i - 1][j - 1] + score : 0,
        );
      }
    }
    let i = m;
    let j = n;
    const lcs: ToolCall[] = [];
    while (i > 0 && j > 0) {
      if (dp[i][j] === dp[i - 1][j]) {
        i -= 1;
      } else if (dp[i][j] === dp[i][j - 1]) {
        j -= 1;
      } else {
        lcs.push(expected[i - 1]);
        i -= 1;
        j -= 1;
      }
    }
    lcs.reverse();
    return { lcs, weightedLength: dp[m][n] };
  }

  private compareDicts(
    d1: Record<string, unknown>,
    d2: Record<string, unknown>,
  ): number {
    if (deepEqual(d1, d2)) return 1;
    if (this.shouldExactMatch) return 0;
    const keys1 = Object.keys(d1);
    const keys2 = Object.keys(d2);
    const matchedKeys = keys1.filter((k) => keys2.includes(k));
    const totalKeys = new Set([...keys1, ...keys2]).size;
    if (totalKeys === 0) return 1;
    let matchScore = 0;
    for (const key of matchedKeys) {
      if (deepEqual(d1[key], d2[key])) {
        matchScore += 1 / totalKeys;
      } else if (
        d1[key] != null &&
        d2[key] != null &&
        typeof d1[key] === "object" &&
        typeof d2[key] === "object" &&
        !Array.isArray(d1[key]) &&
        !Array.isArray(d2[key])
      ) {
        matchScore +=
          this.compareDicts(
            d1[key] as Record<string, unknown>,
            d2[key] as Record<string, unknown>,
          ) / totalKeys;
      }
    }
    return matchScore;
  }

  // --- deterministic tool-calling reason ---

  private generateReason(): string {
    const calledNames = this.toolsCalled.map((t) => t.name);
    const expectedNames = this.expectedTools.map((t) => t.name);

    if (this.shouldExactMatch) {
      const label = this.calculateExactMatchScore()
        ? "Exact match"
        : "Not an exact match";
      return `${label}: expected ${JSON.stringify(expectedNames)}, called ${JSON.stringify(calledNames)}. See details above.`;
    }

    if (this.shouldConsiderOrdering) {
      const { lcs, weightedLength } = this.computeWeightedLcs();
      let score: number;
      if (this.toolsCalled.length === 0 && this.expectedTools.length === 0)
        score = 1;
      else if (this.expectedTools.length === 0) score = 0;
      else score = weightedLength / this.expectedTools.length;

      if (score === 1) {
        return `Correct ordering: all expected tools ${JSON.stringify(expectedNames)} were called in the correct order.`;
      }
      const missing = uniqueMissing(expectedNames, calledNames);
      const outOfOrder = uniqueMissing(
        expectedNames,
        lcs.map((t) => t.name),
      );
      const issues: string[] = [];
      if (missing.length) issues.push(`missing tools ${JSON.stringify(missing)}`);
      if (outOfOrder.length)
        issues.push(`out-of-order tools ${JSON.stringify(outOfOrder)}`);
      return `Incorrect tool usage: ${issues.join(" and ")}; expected ${JSON.stringify(expectedNames)}, called ${JSON.stringify(calledNames)}. See more details above.`;
    }

    if (this.calculateNonExactMatchScore() === 1) {
      return `All expected tools ${JSON.stringify(expectedNames)} were called (order not considered).`;
    }
    const missing = this.expectedTools
      .filter((e) => !this.toolsCalled.some((c) => toolCallEquals(c, e)))
      .map((t) => t.name);
    return `Incomplete tool usage: missing tools ${JSON.stringify(missing)}; expected ${JSON.stringify(expectedNames)}, called ${JSON.stringify(calledNames)}. See more details above.`;
  }

  private constructFinalReason(
    toolCallingReason: string,
    toolSelectionReason: string,
  ): string {
    return (
      "[\n" +
      "\t Tool Calling Reason: " +
      toolCallingReason +
      "\n" +
      "\t Tool Selection Reason: " +
      toolSelectionReason +
      "\n" +
      "]\n"
    );
  }

  isSuccessful(): boolean {
    const ok = this.error == null && (this.score ?? 0) >= this.threshold;
    this.success = ok;
    return ok;
  }

  get name(): string {
    return "Tool Correctness";
  }
}
